NUT-XX: Offline Spilman (unidirectional) channel (#296)
cashubtc/nuts: XX.md

in advance. However, that requires a lot of bandwidth and places heavy strain on the mint.
This NUT describes how a smaller set of proofs, of various denominations, can solve the same
problem, allowing potentially billions of payments with only a few dozen proofs.

# Opening the channel

Alice takes Bob's public key and, with any mint supporting NUT-11, creates a set of 2-of-2 multisig proofs
of various denominations, where both Alice and Bob's signature is needed. Each of these proofs refunds to Alice after a pre-defined time (e.g. one week).
SIG_ALL is used.

## Denominations

_This assumes that 1 millisat is the smallest 'resolution' of interest, but another resolution may be agreed_

Alice shall prepare _2_ such proofs with 1-millisat value.
One of these 1-millisat proofs is a special proof that will be included in every transaction.
Alice shall prepare one proof for every other power of 2: 2-millisats, 4-millisats, 8-millisats, ..., up to some pre-agreed maximum.
The total value of all those proofs is the channel capacity and is the maximum amount that Alice can send to Bob through this channel

> Bob can verify all of these proofs with the mint, before going offline.

Reviewer: Indeed, based on the following LLM summary of `NUT-12`, it seems to be the right tool for the job. I didn't realize this is so involved. I just assumed anyone would be able to verify that proofs are signed by the mint.. Happy to try to understand it together next time we cowork :)

```
Previously, only the mint itself could verify its own signatures. NUT-12 introduces DLEQ proofs, which allow users like Alice (who interacts with the mint) and Carol (who receives ecash from Alice) to verify that the mint (Bob) used the correct private key for signing.

The core idea is to prove that the private key used to create the mint's public key (A) is the same private key used to sign a blinded message (B').
```

The deterministic outputs are created in one of three `context` values:
 - `context = "funding"` for creating the funding token, where each secret is 2-of-2 P2PK with blinded pubkeys
 - `context = "receiver"` for creating the commitment outputs for Charlie, which are P2PK-locked to his per-proof blinded pubkey
 - `context = "sender"` for creating the commitment outputs for Alice, which are P2PK-locked to her per-proof blinded pubkey

The deterministic process is a function of five things:

 - the `channel_id`, the hex representation of the channel_id.
 - the `context` as a string
 - the `amount` in decimal representation, e.g. "32" for 32 satoshis
 - the `index` in decimal representation.
 - the `channel_secret` mentioned earlier. Using this ensures that the mint cannot compute the outputs even if they know the channel_id.

As these are all deterministic and based on information known to both parties,
both parties can construct all these outputs and secrets and blinding factors.

The secret's _nonce_ and the blind-signature _blinding factor_ are computed as follows:

```
deterministic_nonce(...)
 =

>    SHA256(channel_secret || "{channel_id}|{output_context}|{amount}|nonce|{index}")

Author: ```suggestion
   HMAC-SHA256(key = channel_secret, message = "{channel_id}|{output_context}|{amount}|nonce|{index}")
```

blinding_factor(...)
 =

>    SHA256(channel_secret || "{channel_id}|{output_context}|{amount}|blinding|{index}")

Author: ```suggestion
   HMAC-SHA256(key = channel_secret, message = "{channel_id}|{output_context}|{amount}|blinding|{index}")
```

> ```

Reviewer 2: I would advise using an HMAC-SHA256 construction, with channel secret as the key and the rest as the message

Author: Done (in the sense that I have - in the last few seconds - made a suggestion in Github to apply this HMAC-256 construction. Is it what you expected?

- The `"refund"` tag contains Alice's blinded _refund_ pubkey.
- `"n_sigs": "2"` requires both Alice and Charlie's signatures before the NUT-11 `locktime` tag, which is the channel's `expiry_timestamp`
- After that `locktime`, Alice can spend alone using her refund blinded secret key

## Commitment Output Secrets (1-of-1 with Per-Proof Blinded Pubkeys)

Each commitment output is locked to a **unique** blinded pubkey derived from the specific `(amount, index)` of that output.

For the 1-of-1 P2PK secret in the commitment outputs, a per-output
ephemeral private key is computed as:

```
ephemeral_secret = SHA256(
    "Cashu_Spilman_P2BK_ephemeral_v1" ||
    channel_secret ||
    "{channel_id}|{stage2_context}|{amount}|{index}|{retry_counter}"
)
```

where `stage2_context` is `receiver_stage2` or `sender_stage2`, for Charlie's balance and Alice's remainder respectively.
The `retry_counter` is incremented until the hash yields such a valid scalar
for use as the ephemeral secret key.

Then an ECDH shared secret is computed between that ephemeral private key and
the recipient's public key, and the NUT-28 tweak scalar is derived as:

```

> stage2_tweak_scalar = SHA256("Cashu_P2BK_v1" || shared_secret_x || 0x00)

Reviewer 2: I would define `shared_secret_x` at the cost of being redundant. It's better to be exact, so that LLMs can implement more easily.

Author: Done. I've made a suggestion a moment ago to bring clarity to how this scalar is computed.

As this is just the NUT-28 method, I guess we should just refer to that NUT directly, instead of copying-and-pasting the relevant parts of that spec into this. What do you think?

> Then an ECDH shared secret is computed between that ephemeral private key and
> the recipient's public key, and the NUT-28 tweak scalar is derived as:
> 
> ```
> stage2_tweak_scalar = SHA256("Cashu_P2BK_v1" || shared_secret_x || 0x00)
> ```

Author: ```suggestion
 Then an ECDH shared secret is computed as defined by NUT-28 P2BK. Let `Zx` denote the 32-byte x-coordinate shared secret from NUT-28:
 
   - sender side: `Zx = x(eP)` 
   - receiver side: `Zx = x(pE)`

where `e` is the sender ephemeral private key, `E = eG` is the sender ephemeral public key, `p` is the recipient private key, and `P = pG` is the recipient public key.

`stage2_tweak_scalar = SHA256("Cashu_P2BK_v1" || Zx || 0x00)`
      
```

```
for amount in amounts_in_this_keyset:
    if params.maximum_amount_for_one_output > 0 and amount > params.maximum_amount_for_one_output:
        continue

    index = 0
    while True:
        ... attempt to restore with (amount, index) with NUT-09
        if (restore fails for this 'amount' and 'index'):
            break # stop incrementing the index and skip to the next 'amount'
        else:
            .. unblind the signature and add the proof to Alice's wallet
            index += 1
```

This ensures she will get all the outputs and BlindedSignatures which Charlie
created when he did the swap, even if she doesn't know which balance he exited with.

_(Question: Does NUT-09 specify what should happen if the outputs are a mixture of unspent outputs and outputs that were never spent? i.e. Does Alice have to call restore once for each possible output, or can she do a batch call where some will fail and some will succeed?)_

# Charlie's verification

The following information should be sent by Alice to Charlie with, or before, the first payment in this channel,
enabling Charlie to verify everything. He can verify without communicating with the mint.

 - the channel parameters

>  - the funding token, including the DLEQ proofs (NUT-12)

Reviewer 2: With keysets v3, the ecash is issued on the BLS-12-381 curve, which supports special math that allows for public verification. The ecash notes themselves change from being simple algebraic MACs to being proper signatures, so DLEQ is no longer needed.

The nested crypto layer (spending conditions) remains secp-based, so you don't have to change anything.

Author: > so DLEQ is no longer needed.

Indeed. So I guess that (unless v2 keysets are fully replace soon), this NUT will have to say that DLEQ proofs are needed only for v2, and that for v3 the channel partipicants should do the new BLS-based verification instead of checking the (now-missing) DLEQ proofs

Alternatively, Charlie could maintain a per-channel record of the amount of service that
has been provided (image files served, LLM tokens used, kilobytes of data routed by a router ...)
and ensure that each new balance is sufficient to cover the amount due to Charlie for
the service.

Charlie's record of a closed channel should not be deleted until after it has expired,
in order to avoid channel reuse.

If Alice pays using a `channel_id` that is not in either of the two datasets mentioned above,
then Alice is opening a new channel, and Charlie should check everything described above.
As it's a new channel, and assuming the verification is successful, Charlie knows that
the funding token is valid and hasn't already been spent, and he knows that any
commitment transaction that is formed deterministically will also have unspent outputs.

# Timing

To improve privacy further, when the channel is closed and the first of the two stages
is executed, both parties should delay the second stage.
This delay is to make it more difficult for the mint to correlate the three 'exit swaps' with each other.

# proof-of-concept

> _As of 2026-03-20, the implementation includes P2BK (Pay-to-Blinded-Key) with per-proof blinding for stage 2 outputs. The funding and commitment outputs are deterministic with blinded pubkeys. Fees are taken into account fully. The token-state-check and token-restore endpoints are used. The implementation is [here](https://github.com/SatsAndSports/demo_of_spillman_cashu_channel/tree/spilman.channel/crates/cdk/src/spilman), with many integration tests._

Reviewer 2: how can per-proof P2BK blinding work with SIG_ALL? IIRC this was a known problem, where SIG_ALL requires all inputs outputs to be signed together by the pubkey(s) and all inputs having the same spending conditions.

Author: There are two stages in this, each of which involves blinding

The 'channel token' itself can only be spent with SIG_ALL and therefore, as you say, the individual proofs must be identical in certain respects. There are two keys for the 2-of-2 spend path, and a third for the post-expiry refund. Those three keys are different from each other, and different from the user's "real" pubkey, but they are identical across all the outputs in this 'channel token'

The second stage is where the 'channel token' is spent and it creates outputs to distribute the funds between the two parties. The proofs created from those outputs are themselves 1-of-1 P2BK, and they are _not_ SIG_ALL, and therefore each output is blinded differently
